qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py

```diff
@@ -0,0 +1,151 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2025 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+from instrumentview.Detectors import DetectorInfo
+import instrumentview.Projections.SphericalProjection as iv_spherical
+import instrumentview.Projections.CylindricalProjection as iv_cylindrical
+
+from mantid.dataobjects import Workspace2D
+import numpy as np
+import math
+
+
+class FullInstrumentViewModel:
+    """Model for the Instrument View Window. Will calculate detector positions, indices, and integrated counts that give the colours"""
+
+    _sample_position = np.array([0, 0, 0])
+    _source_position = np.array([0, 0, 0])
+    _invalid_index = -1
+    _data_min = 0.0
+    _data_max = 0.0
+
+    def __init__(self, workspace: Workspace2D):
+        """For the given workspace, calculate detector positions, the map from detector indices to workspace indices, and integrated
+        counts. Optionally will draw detector geometry, e.g. rectangular bank or tube instead of points."""
+        self._workspace = workspace
+
+    def setup(self):
+        self._detector_info = self._workspace.detectorInfo()
+        self._component_info = self._workspace.componentInfo()
+        self._sample_position = np.array(self._component_info.samplePosition()) if self._component_info.hasSample() else np.zeros(3)
+        has_source = self._workspace.getInstrument().getSource() is not None
+        self._source_position = np.array(self._component_info.sourcePosition()) if has_source else np.array([0, 0, 0])
+        self._detector_ids = np.array(self._detector_info.detectorIDs())
+        self._detector_positions = np.array([self._detector_info.position(i) for i in range(len(self._detector_ids))])
+        self._spherical_positions = np.array([self._detector_info.position(i).getSpherical() for i in range(len(self._detector_ids))])
```

Comment:
Is it possible (maybe in a different PR) to keep `self._detector_positions` as a V3D in the line above, then call `.getSpherical()` before converting to numpy array (avoid getting it twice) - might be a bottleneck?

Reply:
Also consider using `np.asarray`  - I think this sometimes avoids copying when making the array - not sure if this will be possible for V3D?

Reply:
The other alternative for now to get a quick speed up is to call `CreateDetectorTable` algorithm (with new-ish option `IncludeDetectorPosition=True`) then you can get the spherical  like so `tbl.column('Position')` - which returns a list of V3Ds on which you can call `getSpherical`